### ISSUE TITLE
--store with remote store unnecessarily uploads direct path references each time

### ISSUE THREAD

**Issue by LisannaAtHome** [NixOS/nix#2272]
Every time I use the `--store` flag of nix-build with a remote store (`ssh-ng://`), it seems to have to upload all my direct path references to the remote every time I run the command, even if the direct path references haven't changed.

This makes using remote stores from a location with weak upload bandwidth really suck.

**Comment by edolstra** [comment]
Can you provide an example? It's not clear why Nixpkgs would have to be uploaded to the remote at all.

**Comment by ledettwy** [comment]
Doing some more digging on this, and I was mistaken in saying that it's uploading nixpkgs to the remote. nixpkgs is about 80MB, and it's uploading 20MB.

What's more likely happening here is that it's uploading directories whose paths I reference directly in the expression (the sizes add up). But, again, these directories aren't changing between invocations of nix-build, but are still re-uploaded each time. Why wouldn't these be being uploaded to the same location each time (thus allowing nix-build to skip uploading if the path in the remote store already exists from a previous invocation)? Is that not how direct path references work?

**Comment by edolstra** [comment]
Yes, `RemoteStore::addToStore()` doesn't check whether a path already exists in the remote store. It just sends it over and lets the remote side decide what to do with it.

**Comment by ledettwy** [comment]
OK, that explains the behavior. I guess `remoteStore::addToStore()` is being called on all of the build-time dependencies, which causes them to get re-uploaded each time unnecessarily.

This seems less than efficient - are there any technical reasons for this implementation?

**Comment by edolstra** [comment]
Yes, it's because `RemoteStore` (unlike what the name suggests) is written for the case where the Nix store is on the same machine (i.e. using the Nix daemon). In that case, having the client check whether the path already exists is (probably) less efficient that having the daemon do it.

**Comment by ledettwy** [comment]
Ah, okay. I guess that's a bit unfortunate. Would be cool if it had a different behavior for when a remote store is specified. This would speed up more than just `nix-build --store` usecases, since I think other stuff like `nixos-rebuild --build-host` and distributed building also all probably call `RemoteStore::addToStore()` on the build dependencies.

**Comment by ledettwy** [comment]
Nevermind, remote/distributed building doesn't need to reupload the build dependencies each time. Does it use a different mechanism?

**Comment by lheckemann** [comment]
>Does it use a different mechanism?

Yes. The local store is still used as the "base of operations", and the remote builders are only invoked when a derivation goal is reached (at which point the paths referred to directly have already been imported to the local store). The paths are then copied to the remote builders as needed (or not needed!).
